Allow listing only non-deleted kv secrets

Shaeli opened the pull request:
# Overview

This allows listing Vault KV secrets while filtering out deleted entries. The default behavior of listing secrets hasn't changed, but a new boolean parameter named `exclude_deleted` is added, that defaults to false.
When this boolean is set to true,  secrets where the current version is deleted will be filtered out from the `list` output.

## Motivation

Currently, when looping through Vault secrets from remote locations (e.g., Australia), the process can be slow because soft-deleted secrets appear in the secret list operations and require additional metadata checks to determine their status.  This creates an incentive for users to perform destructive metadata deletions just to clean up listings, even when there's a good case for keeping deleted secrets around for potential rollback scenarios.


This change allows users to exclude soft deleted entries when listing secrets to speed up Vault searches and avoid having to destroy data unnecessarily


## Changes

- Added optional `exclude_deleted` boolean parameter to metadata list operations
- When `exclude_deleted=true`, filters out keys where the current version has been soft-deleted
- Backward compatibility: default behavior unchanged
- Only checks current version deletion status (not historical versions)

If this change is accepted, I can update the documentation on `hashicorp/vault` side as well.

I'm not super familiar with Golang, so please let me know if anything needs to be modified / can be improved ☺️

# Related Issues/Pull Requests
Solves: https://github.com/hashicorp/vault/issues/8432#issuecomment-1979747487

# Contributor Checklist
[ ] Add relevant docs to upstream Vault repository, or sufficient reasoning why docs won’t be added yet
[My Docs PR Link](link)
[Example](https://github.com/hashicorp/vault/commit/2715f5cec982aabc7b7a6ae878c547f6f475bba6)
[ ] Add output for any tests not ran in CI to the PR description (eg, acceptance tests)
[ ] Backwards compatible

# PCI review checklist
<!-- heimdall_github_prtemplate:grc-pci_dss-2024-01-05 -->
- [x] I have documented a clear reason for, and description of, the change I am making.
- [ ] If applicable, I've documented a plan to revert these changes if they require more than reverting the pull request.
- [ ] If applicable, I've documented the impact of any changes to security controls.
Examples of changes to security controls include using new access control methods, adding or removing logging pipelines, etc.

robmonte commented on path_metadata_test.go:
```diff
@@ -1560,3 +1560,379 @@ func TestVersionedKV_Metadata_Patch_NilsUnset(t *testing.T) {
+
+	keys = resp.Data["keys"].([]string)
+	// Should include: bar, dir/, foo (directories are always included)
+	if len(keys) != 3 || keys[0] != "bar" || keys[1] != "dir/" || keys[2] != "foo" {
+		t.Fatalf("expected keys [bar, dir/, foo] (with directory), got %v", keys)
```
This step of the test is failing for me but I don't have time to look further into it unfortunately.

```
=== RUN   TestVersionedKV_Metadata_List_ExcludeDeleted
    /Users/robmonte/github/vault-plugin-secrets-kv/path_metadata_test.go:1814: expected keys [bar, dir/, foo] (with directory), got [bar baz dir/]
--- FAIL: TestVersionedKV_Metadata_List_ExcludeDeleted (0.00s)
```

Shaeli replied:
Oops, I simplified the tests just before pushing and forgot to update this one, it's actually expected to fail in this current state!  Thanks for noticing this. I updated it :)

robmonte reviewed (commented):
Great PR, thank you! Just a couple comments, and we really appreciate your excellent test coverage.

Shaeli commented:
@robmonte Thanks for the review as I'm not very familiar with Golang. I updated the PR with your suggestions and fixed the failing test, let me know if anything else needs to be modified!

robmonte commented:
Apologies, I was about to merge but remembered that I forgot to mention one thing. 😅
Could you please add a new feature line and description for this field under the "unreleased" section of the CHANGELOG.md file? It will help us when its time for a new release. Thanks!

robmonte commented:
@Shaeli So our docs process is changing currently and is based on https://github.com/hashicorp/web-unified-docs now instead of the website directory within Vault.

If you could open a PR there targeting the 1.21.x branch and link it to me here, I can make sure it gets reviewed and merged to go with this feature.

As for releasing together, since we provide update builds on a regular cadence (in this case when Vault 1.21 goes GA), we simply ensure all our documentation updates are ready to go live by the time Vault releases. This code change can go in now while still waiting for the docs to be finalized.
